jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java

```diff
@@ -129,25 +111,72 @@ private static native void checkpointJVMImpl(String imageDir,
 	 */
 	public CRIUSupport(Path imageDir) {
 		/* [IF JAVA_SPEC_VERSION < 17] */
+		@SuppressWarnings("removal")
+		/* [ENDIF] JAVA_SPEC_VERSION < 17 */
 		SecurityManager manager = System.getSecurityManager();
 		if (manager != null) {
 			manager.checkPermission(CRIU_DUMP_PERMISSION);
 		}
-		/* [ENDIF] JAVA_SPEC_VERSION < 17 */
 
 		setImageDir(imageDir);
 	}
 
+	@SuppressWarnings("restriction")
+	private synchronized static void init() {
+		if (!initComplete) {
+			if (loadNativeLibrary()) {
+				criuSupportEnabled = isCRIUSupportEnabledImpl();
+
+				try {
+					Field f = Unsafe.class.getDeclaredField("theUnsafe"); //$NON-NLS-1$
+					f.setAccessible(true);
+					unsafe = (Unsafe) f.get(null);
+				} catch (NoSuchFieldException | SecurityException | IllegalArgumentException | IllegalAccessException e) {
+					throw new InternalError(e);
+				}
+			}
+
+			initComplete = true;
+		}
+	}
+
 	/**
 	 * Queries if CRIU support is enabled.
 	 *
 	 * @return TRUE is support is enabled, FALSE otherwise
 	 */
 	public static boolean isCRIUSupportEnabled() {
+		if (!initComplete) {
```

Comment:
It would be more straightforward to just make this method and the native library loading one `synchronized` - creating a checkpoint isn't going to be a performance critical operation so we shouldn't worry about the cost of a synchronized block here.

Better to be simple and obviously correct than clever and potentially racy